# 전처리 & 머신러닝 통합 개인과제

## Part 1. 고객 데이터 품질 개선

### 실무 시나리오

전자상거래 기업의 CRM팀은 고객별 구매 패턴을 분석하고 구매 가능성이 높은 고객을 선별하려고 합니다.  
그러나 전달받은 원본 고객 데이터에는 잘못된 자료형, 범주 표기 불일치, 결측치, 중복 데이터와 극단값이 포함되어 있어 바로 분석에 사용할 수 없습니다.

CRM팀은 **데이터 분석 담당자**로서 원본 데이터의 품질 문제를 진단하고, 이후 Part 2의 EDA와 Part 3의 구매 예측 모델링에 사용할 수 있는 분석용 데이터셋을 만들어야 합니다.

### 과제 목표

- 데이터의 구조·자료형·기초 분포를 확인하고 주요 품질 문제를 파악합니다.
- 결측치, 중복값, 이상치와 범주 표기 불일치를 분석 목적에 맞게 처리합니다.
- 조건 기반 데이터 추출과 파생변수 생성을 수행합니다.
- 전처리 결과를 검증하고 다음 Part에서 사용할 CSV 파일로 저장합니다.

### 사용 환경 및 데이터

- Python 3.X
- pandas, numpy
- `전자상거래_고객구매_원본데이터.csv`

### 제출 결과물

- Part 1 실습 노트북
- `전자상거래_고객구매_전처리완료.csv`

> 문제 1~3은 필수 문제입니다.

## 문제 1. 원본 고객 데이터 품질 진단

### 업무 상황

CRM팀에 분석 일정을 공유하기 전에 원본 데이터가 실제 분석에 사용할 수 있는 상태인지 확인해야 합니다.  
데이터를 수정하기 전에 구조와 품질 문제를 먼저 점검하고, 이후 처리해야 할 항목을 정리하세요.

### 요구사항

1. 원본 데이터를 불러오고 데이터의 크기, 컬럼, 자료형과 기초 통계량을 확인하세요.
2. 컬럼별 결측치와 전체 행 기준 중복 데이터를 확인하세요.
3. 주요 범주형 컬럼의 값을 확인하여 표기 불일치 여부를 파악하세요.
4. 이후 정제가 필요한 주요 품질 문제를 간단히 정리하세요.

### 힌트

- 데이터 구조와 기초 통계량을 함께 확인하면 자료형 오류와 비정상 범위를 찾기 쉽습니다.
- 범주형 컬럼은 고유값을 확인하여 대소문자, 공백, 한글·영문 혼용 여부를 살펴볼 수 있습니다.

In [1]:
import pandas as pd
import numpy as np
from pathlib import Path

DATA_PATH = Path("전자상거래_고객구매_원본데이터.csv")

# TODO: 원본 데이터를 불러와 df에 저장하세요.
df = pd.read_csv("전자상거래_고객구매_데이터.csv")

# TODO: 데이터의 크기, 컬럼, 자료형과 기초 통계량을 확인하세요.
df.head()
df.shape
df.columns
df.dtypes
df.describe(include="all")
# TODO: 결측치, 중복 데이터와 주요 범주형 컬럼의 값을 확인하세요.
df.isna().sum()
df.duplicated().sum()

np.int64(60)

In [2]:
duplicate_rows = df.duplicated(keep=False)
df[duplicate_rows].head(10)

,CustomerID,Gender,Age,Region,MembershipLevel,VisitCount,AveragePurchaseAmount,TotalPurchaseAmount,SatisfactionScore,LastPurchaseDate,CouponUsed,PreferredCategory,SignupDate,Email,PurchaseStatus
10,CUST100037,Female,53,GYEONGGI,Basic,10,153900.0,629800.0,4.4,2026-02-18,N,Electronics,2024-09-05,customer37@example.com,1
13,CUST100375,Male,34,Daejeon,Basic,13,148300.0,609500.0,3.3,2026-04-24,Y,Fashion,2021-08-28,customer375@example.com,0
24,CUST100883,female,43,Gyeonggi,Basic,7,113600.0,360500.0,3.9,2026/01/15,Y,Sports,2022-02-13,customer883@example.com,0
29,CUST100148,Male,57,Incheon,Silver,9,99600.0,207200.0,2.9,2026-05-26,N,Food,2021-04-07,customer148@example.com,0
31,CUST100419,Female,20,Seoul,Basic,6,102300.0,225100.0,2.2,2026-06-11,N,Fashion,2023-08-19,customer419@example.com,0
65,CUST100865,Other,NaN,Gyeonggi,VIP,9,209200.0,445200.0,3.4,2026-04-08,Y,Sports,2025-03-17,customer865@example.com,1
69,CUST101379,Male,62,Busan,Basic,7,45100.0,97400.0,4.2,2026-04-19,N,Sports,2025-06-24,customer1379@example.com,0
91,CUST100419,Female,20,Seoul,Basic,6,102300.0,225100.0,2.2,2026-06-11,N,Fashion,2023-08-19,customer419@example.com,0
105,CUST100014,Male,50,Daegu,Silver,12,162300.0,21000000.0,4.1,2026-03-06,Y,Electronics,2025-12-21,customer14@example.com,1
106,CUST100046,Male,26,Gwangju,Silver,10,150400.0,460200.0,4.7,2026-05-07,Y,Fashion,2022-07-14,customer46@example.com,1


In [3]:
df["CustomerID"].duplicated().sum()

np.int64(60)

In [4]:
df["Gender"].unique()

<StringArray>
[  'Female',     'Male',   'female',    'Other', ' FEMALE ',     'male',
        'F',        nan,   ' MALE ',        'M']
Length: 10, dtype: str

In [5]:
df["Gender"].value_counts(dropna=False)

Gender
Female      713
Male        682
Other        50
NaN          26
female       19
 FEMALE      15
male         14
F            14
 MALE        14
M            13
Name: count, dtype: int64

In [6]:
df["Region"].unique()

<StringArray>
[         nan, ' GYEONGGI ',    'Gwangju',      'Daegu',   'Gyeonggi',
    'Daejeon',      'Busan',      'Seoul',    'Incheon',      'seoul',
    ' Seoul ',   'gyeonggi',      'busan',    ' Busan ']
Length: 14, dtype: str

In [7]:
df["Region"].value_counts(dropna=False)


Region
Seoul         401
Gyeonggi      362
Daejeon       172
Gwangju       155
Busan         138
Incheon       132
Daegu         114
NaN            31
 GYEONGGI      12
seoul          12
 Seoul         12
gyeonggi       11
busan           4
 Busan          4
Name: count, dtype: int64

In [8]:
df["MembershipLevel"].unique()

<StringArray>
[ 'silver',   'Basic',  'Silver',     'VIP',    'Gold',    'gold',       nan,
 ' Basic ',  'SILVER',  ' Gold ',   'basic']
Length: 11, dtype: str

In [9]:
df["MembershipLevel"].value_counts(dropna=False)

MembershipLevel
Basic      659
Silver     413
Gold       267
VIP        115
NaN         22
 Basic      21
basic       21
silver      14
SILVER      12
gold         8
 Gold        8
Name: count, dtype: int64

In [10]:
df["CouponUsed"].unique()

<StringArray>
['N', ' n ', 'Y', 'No', 'Yes', ' y ']
Length: 6, dtype: str

In [11]:
df["CouponUsed"].value_counts(dropna=False)

CouponUsed
N      840
Y      628
No      28
 n      25
Yes     20
 y      19
Name: count, dtype: int64

In [12]:
df["PreferredCategory"].unique()

<StringArray>
[       'Beauty',          'Home',       'Fashion',          'Food',
   'Electronics',        'beauty',        'Sports',      ' Beauty ',
       'fashion',       'FASHION', ' Electronics ',             nan,
   'electronics']
Length: 13, dtype: str

In [13]:
df["PreferredCategory"].value_counts(dropna=False)

PreferredCategory
Fashion          343
Electronics      257
Food             255
Beauty           237
Home             218
Sports           175
NaN               26
fashion           10
FASHION           10
electronics        8
beauty             7
 Beauty            7
 Electronics       7
Name: count, dtype: int64

In [14]:
df_clean = df.copy()
print("원본 크기:", df.shape)
print("정제용 크기:", df_clean.shape)

원본 크기: (1560, 15)
정제용 크기: (1560, 15)


In [15]:
df['Age'] = df['Age'].str.strip()

df.dropna(axis=0, inplace=True)

In [16]:
df_clean["Age"].value_counts(dropna=False)

Age
47         61
38         56
37         54
41         54
46         53
           ..
101         1
36세         1
thirty      1
67          1
unknown     1
Name: count, Length: 79, dtype: int64

In [17]:
df_clean["Age"] = df_clean["Age"].astype("string")

In [18]:
df_clean["Age"] = df_clean["Age"].str.replace("세", "", regex=False)

In [19]:
df_clean["Age"].value_counts(dropna=False)

Age
47         61
38         56
37         54
41         54
46         53
           ..
미입력         1
101         1
thirty      1
67          1
unknown     1
Name: count, Length: 69, dtype: Int64

In [20]:
df_clean["Age"] = df_clean["Age"].str.replace(",", "", regex=False)

In [21]:
df_clean["Age"].value_counts(dropna=False)

Age
47         61
38         56
37         54
41         54
46         53
           ..
미입력         1
101         1
thirty      1
67          1
unknown     1
Name: count, Length: 69, dtype: Int64

In [22]:
df_clean["Age"] = df_clean["Age"].str.strip()

In [23]:
df_clean["Age"].value_counts(dropna=False)

Age
47         61
38         56
37         54
41         54
46         53
           ..
미입력         1
101         1
thirty      1
67          1
unknown     1
Name: count, Length: 69, dtype: Int64

In [24]:
df_clean["Age"].dtype

<StringDtype(storage='python', na_value=<NA>)>

In [25]:
original_present = df["Age"].notna()
converted_missing = df_clean["Age"].isna()
failed = original_present & converted_missing
df.loc[failed, "Age"].value_counts(dropna=False)

Series([], Name: count, dtype: int64)

In [26]:
df_clean["AveragePurchaseAmount"].value_counts(dropna=False)

AveragePurchaseAmount
NaN         45
5000.0      13
143400.0     7
106100.0     6
148300.0     5
            ..
138700.0     1
172700.0     1
134600.0     1
113100.0     1
73500.0      1
Name: count, Length: 1023, dtype: int64

In [27]:
df_clean["AveragePurchaseAmount"] = df_clean["AveragePurchaseAmount"].astype("string")

In [28]:
df_clean["AveragePurchaseAmount"] = df_clean["AveragePurchaseAmount"].str.replace(",", "", regex=False)

In [29]:
df_clean["AveragePurchaseAmount"].value_counts(dropna=False)

AveragePurchaseAmount
<NA>        45
5000.0      13
143400.0     7
106100.0     6
148300.0     5
            ..
138700.0     1
172700.0     1
134600.0     1
113100.0     1
73500.0      1
Name: count, Length: 1023, dtype: Int64

In [30]:
df_clean["AveragePurchaseAmount"] = pd.to_numeric(df_clean["AveragePurchaseAmount"], errors="coerce")

In [31]:
df_clean["AveragePurchaseAmount"].value_counts(dropna=False)

AveragePurchaseAmount
<NA>        45
5000.0      13
143400.0     7
106100.0     6
148300.0     5
            ..
138700.0     1
172700.0     1
134600.0     1
113100.0     1
73500.0      1
Name: count, Length: 1015, dtype: Int64

In [32]:
df_clean["AveragePurchaseAmount"].dtype

Float64Dtype()

In [33]:
original_present = df["AveragePurchaseAmount"].notna()
converted_missing = df_clean["AveragePurchaseAmount"].isna()
failed = original_present & converted_missing
df.loc[failed, "AveragePurchaseAmount"].value_counts(dropna=False)

Series([], Name: count, dtype: int64)

In [34]:
df_clean[["Age", "AveragePurchaseAmount"]].describe()

,AveragePurchaseAmount
count,1515.0
mean,120540.594059
std,52777.455516
min,5000.0
25%,83300.0
50%,117500.0
75%,152150.0
max,292600.0


In [35]:
df_clean["LastPurchaseDate"].head(10)

0    2026-02-07
1    2026-04-22
2    2026-06-06
3    2026-05-07
4    2026-04-16
5           NaN
6    2026-01-31
7    2026-01-21
8    2026-06-12
9    2026-03-19
Name: LastPurchaseDate, dtype: str

In [36]:
df_clean["LastPurchaseDate"] = pd.to_datetime(df_clean["LastPurchaseDate"], errors="coerce", format="mixed")

In [37]:
df_clean["LastPurchaseDate"].head(10)

0   2026-02-07
1   2026-04-22
2   2026-06-06
3   2026-05-07
4   2026-04-16
5          NaT
6   2026-01-31
7   2026-01-21
8   2026-06-12
9   2026-03-19
Name: LastPurchaseDate, dtype: datetime64[us]

In [38]:
df_clean["LastPurchaseDate"].isna().sum()

np.int64(30)

In [39]:
original_present = df["LastPurchaseDate"].notna()
converted_missing = df_clean["LastPurchaseDate"].isna()
failed = original_present & converted_missing
df.loc[failed, "LastPurchaseDate"].value_counts(dropna=False)

Series([], Name: count, dtype: int64)

In [40]:
df_clean["SignupDate"].head(10)

0    2022-12-03
1    2021-07-16
2    2022-02-15
3    2021-10-27
4    2025-11-08
5    2024-01-05
6    2025-01-28
7    2021-07-01
8    2024-05-03
9    2023-10-13
Name: SignupDate, dtype: str

In [41]:
df_clean["SignupDate"] = pd.to_datetime(df_clean["SignupDate"], errors="coerce", format="mixed")

In [42]:
df_clean["SignupDate"].head(10)

0   2022-12-03
1   2021-07-16
2   2022-02-15
3   2021-10-27
4   2025-11-08
5   2024-01-05
6   2025-01-28
7   2021-07-01
8   2024-05-03
9   2023-10-13
Name: SignupDate, dtype: datetime64[us]

In [43]:
df_clean["SignupDate"].isna().sum()

np.int64(0)

In [44]:
original_present = df["SignupDate"].notna()
converted_missing = df_clean["SignupDate"].isna()
failed = original_present & converted_missing
df.loc[failed, "SignupDate"].value_counts(dropna=False)

Series([], Name: count, dtype: int64)

In [45]:
df_clean["Gender"].value_counts(dropna=False)

Gender
Female      713
Male        682
Other        50
NaN          26
female       19
 FEMALE      15
male         14
F            14
 MALE        14
M            13
Name: count, dtype: int64

In [46]:
df_clean["Gender"] = df_clean["Gender"].astype("string")

In [47]:
df_clean["Gender"] = df_clean["Gender"].str.strip()

In [48]:
df_clean["Gender"].value_counts(dropna=False)

Gender
Female    713
Male      682
Other      50
<NA>       26
female     19
FEMALE     15
male       14
F          14
MALE       14
M          13
Name: count, dtype: Int64

In [49]:
df_clean["Gender"] = df_clean["Gender"].str.lower()

In [50]:
df_clean["Gender"].value_counts(dropna=False)

Gender
female    747
male      710
other      50
<NA>       26
f          14
m          13
Name: count, dtype: Int64

In [51]:
gender_map = {
    "female": "Female",
    "f": "Female",
    "male": "Male",
    "m": "Male",
    "other": "Other"
}

In [52]:
df_clean["Gender"] = df_clean["Gender"].map(gender_map)

In [53]:
df_clean["Gender"].value_counts(dropna=False)

Gender
Female    761
Male      723
Other      50
NaN        26
Name: count, dtype: int64

In [54]:
df_clean["Region"].value_counts(dropna=False)

Region
Seoul         401
Gyeonggi      362
Daejeon       172
Gwangju       155
Busan         138
Incheon       132
Daegu         114
NaN            31
 GYEONGGI      12
seoul          12
 Seoul         12
gyeonggi       11
busan           4
 Busan          4
Name: count, dtype: int64

In [55]:
df_clean["Region"] = df_clean["Region"].astype("string")

In [56]:
df_clean["Region"] = df_clean["Region"].str.strip()

In [57]:
df_clean["Region"].value_counts(dropna=False)

Region
Seoul       413
Gyeonggi    362
Daejeon     172
Gwangju     155
Busan       142
Incheon     132
Daegu       114
<NA>         31
GYEONGGI     12
seoul        12
gyeonggi     11
busan         4
Name: count, dtype: Int64

In [58]:
df_clean["Region"] = df_clean["Region"].str.lower()

In [59]:
df_clean["Region"].value_counts(dropna=False)

Region
seoul       425
gyeonggi    385
daejeon     172
gwangju     155
busan       146
incheon     132
daegu       114
<NA>         31
Name: count, dtype: Int64

In [60]:
df_clean["Region"] = df_clean["Region"].str.title()

In [61]:
df_clean["Region"].value_counts(dropna=False)

Region
Seoul       425
Gyeonggi    385
Daejeon     172
Gwangju     155
Busan       146
Incheon     132
Daegu       114
<NA>         31
Name: count, dtype: Int64

In [62]:
df_clean["MembershipLevel"].value_counts(dropna=False)

MembershipLevel
Basic      659
Silver     413
Gold       267
VIP        115
NaN         22
 Basic      21
basic       21
silver      14
SILVER      12
gold         8
 Gold        8
Name: count, dtype: int64

In [63]:
df_clean["MembershipLevel"] = df_clean["MembershipLevel"].astype("string")

In [64]:
df_clean["MembershipLevel"] = df_clean["MembershipLevel"].astype("string")

In [65]:
df_clean["MembershipLevel"] = df_clean["MembershipLevel"].str.strip()

In [66]:
df_clean["MembershipLevel"].value_counts(dropna=False)

MembershipLevel
Basic     680
Silver    413
Gold      275
VIP       115
<NA>       22
basic      21
silver     14
SILVER     12
gold        8
Name: count, dtype: Int64

In [67]:
df_clean["MembershipLevel"] = df_clean["MembershipLevel"].str.lower()

In [68]:
df_clean["MembershipLevel"].value_counts(dropna=False)

MembershipLevel
basic     701
silver    439
gold      283
vip       115
<NA>       22
Name: count, dtype: Int64

In [69]:
df_clean["MembershipLevel"] = df_clean["MembershipLevel"].str.title()

In [70]:
df_clean["MembershipLevel"].value_counts(dropna=False)

MembershipLevel
Basic     701
Silver    439
Gold      283
Vip       115
<NA>       22
Name: count, dtype: Int64

In [71]:
df_clean["CouponUsed"].value_counts(dropna=False)

CouponUsed
N      840
Y      628
No      28
 n      25
Yes     20
 y      19
Name: count, dtype: int64

In [72]:
df_clean["CouponUsed"] = df_clean["CouponUsed"].astype("string")

In [73]:
df_clean["CouponUsed"] = df_clean["CouponUsed"].str.strip()

In [74]:
df_clean["CouponUsed"].value_counts(dropna=False)

CouponUsed
N      840
Y      628
No      28
n       25
Yes     20
y       19
Name: count, dtype: Int64

In [75]:
coupon_map = {"y": "Y", "yes": "Y", "n": "N", "no": "N"}

In [76]:
df_clean["CouponUsed"] = df_clean["CouponUsed"].map(coupon_map)

In [77]:
df_clean["CouponUsed"].value_counts(dropna=False)

CouponUsed
NaN    1516
N        25
Y        19
Name: count, dtype: int64

In [78]:
df_clean["PreferredCategory"].value_counts(dropna=False)

PreferredCategory
Fashion          343
Electronics      257
Food             255
Beauty           237
Home             218
Sports           175
NaN               26
fashion           10
FASHION           10
electronics        8
beauty             7
 Beauty            7
 Electronics       7
Name: count, dtype: int64

In [79]:
df_clean["PreferredCategory"] = df_clean["PreferredCategory"].astype("string")

In [80]:
df_clean["PreferredCategory"] = df_clean["PreferredCategory"].str.strip()

In [81]:
df_clean["PreferredCategory"].value_counts(dropna=False)

PreferredCategory
Fashion        343
Electronics    264
Food           255
Beauty         244
Home           218
Sports         175
<NA>            26
fashion         10
FASHION         10
electronics      8
beauty           7
Name: count, dtype: Int64

In [82]:
df_clean["PreferredCategory"] = df_clean["PreferredCategory"].str.lower()

In [83]:
df_clean["PreferredCategory"].value_counts(dropna=False)

PreferredCategory
fashion        363
electronics    272
food           255
beauty         251
home           218
sports         175
<NA>            26
Name: count, dtype: Int64

In [84]:
df_clean["PreferredCategory"] = df_clean["PreferredCategory"].str.title()

In [85]:
df_clean["PreferredCategory"].value_counts(dropna=False)

PreferredCategory
Fashion        363
Electronics    272
Food           255
Beauty         251
Home           218
Sports         175
<NA>            26
Name: count, dtype: Int64

In [86]:
df_clean.dtypes

CustomerID                          str
Gender                              str
Age                              string
Region                           string
MembershipLevel                  string
VisitCount                        int64
AveragePurchaseAmount           Float64
TotalPurchaseAmount             float64
SatisfactionScore               float64
LastPurchaseDate         datetime64[us]
CouponUsed                          str
PreferredCategory                string
SignupDate               datetime64[us]
Email                               str
PurchaseStatus                    int64
dtype: object

### 품질 진단 결과

- 자료형 정리가 필요한 컬럼: Age, AveragePurchaseAmount
- 표기 통일이 필요한 컬럼: Gender, Age, LastPurchaseDate, CouponUsed
- 결측치가 있는 컬럼: Gender, Age, Region, MembershipLev,el, AveragePurchaseAmount, SatisfactionScore,
LastPurchaseDate, PreferredCategory
- 중복 데이터 확인 결과: 중복 데이터 개수 60개 존재
- 이상치 확인이 필요한 컬럼: Age, VisitCount, TotalPurchaseAmount
- 이후 처리할 주요 항목: 

## 문제 2. 분석 가능한 고객 데이터로 정제

### 업무 상황

품질 진단 결과를 바탕으로 고객 데이터를 분석 가능한 상태로 정리해야 합니다.  
처리 과정에서 고객의 실제 구매 행동 정보가 불필요하게 손실되지 않도록 데이터 특성을 고려하세요.

### 요구사항

1. 원본 데이터를 보존한 상태에서 정제용 데이터를 생성하세요.
2. 분석에 맞지 않는 수치형·날짜형 자료형과 범주 표기를 정리하세요.
3. 결측치와 전체 행 기준 중복 데이터를 처리하세요.
4. 주요 수치형 컬럼의 이상치를 탐지하고 적절한 방법으로 처리하세요.
5. 처리 전후의 결측치, 중복값과 이상치 상태를 확인하세요.

### 힌트

- 변환할 수 없는 문자열은 결측치로 바꾼 뒤 일관되게 처리할 수 있습니다.
- 이상치는 무조건 삭제하기보다 실제 우수 고객의 행동일 가능성도 고려하세요.
- IQR은 이상치 후보를 확인하는 대표적인 방법입니다.

In [ ]:
# TODO: 원본을 보존하고 정제용 데이터 df_clean을 생성하세요.
df_clean = df.copy()
print("원본 크기:", df.shape)
print("정제용 크기:", df_clean.shape)

# TODO: 수치형·날짜형 자료형과 범주 표기를 정리하세요.
df['Age'] = df['Age'].str.strip()
df.dropna(axis=0, inplace=True)
df_clean["Age"].value_counts(dropna=False)



원본 크기: (1317, 15)
정제용 크기: (1317, 15)


Age
47         51
41         50
38         49
32         49
46         48
           ..
36세         1
thirty      1
67          1
64          1
unknown     1
Name: count, Length: 75, dtype: int64

In [ ]:
df_clean["Age"] = df_clean["Age"].str.replace(",", "", regex=False)

In [90]:
df_clean["Age"].value_counts(dropna=False)

Age
47         51
41         50
38         49
32         49
46         48
           ..
36세         1
thirty      1
67          1
64          1
unknown     1
Name: count, Length: 75, dtype: int64

In [91]:
df_clean["Age"] = df_clean["Age"].str.replace("세", "", regex=False)

In [92]:
df_clean["Age"].value_counts(dropna=False)

Age
47         51
41         50
38         49
32         49
46         48
           ..
68          1
thirty      1
67          1
64          1
unknown     1
Name: count, Length: 67, dtype: int64

In [ ]:
df_clean["Age"] = df_clean["Age"].str.strip()


In [94]:
df_clean["Age"].value_counts(dropna=False)

Age
47         51
41         50
38         49
32         49
46         48
           ..
68          1
thirty      1
67          1
64          1
unknown     1
Name: count, Length: 67, dtype: int64

In [95]:
df_clean["Age"] = pd.to_numeric(df_clean["Age"], errors="coerce")

In [96]:
df_clean["Age"].value_counts(dropna=False)

Age
47.0     51
41.0     50
38.0     49
32.0     49
46.0     48
         ..
99.0      1
101.0     1
68.0      1
67.0      1
64.0      1
Name: count, Length: 63, dtype: int64

In [97]:
df_clean["Age"].dtype

dtype('float64')

In [98]:
original_present = df["Age"].notna()
converted_missing = df_clean["Age"].isna()
failed = original_present & converted_missing
df.loc[failed, "Age"].value_counts(dropna=False)

Age
-          1
?          1
미입력        1
thirty     1
unknown    1
Name: count, dtype: int64

In [99]:
df_clean["AveragePurchaseAmount"].value_counts(dropna=False)

AveragePurchaseAmount
5000.0      9
143400.0    7
148300.0    5
99600.0     5
116900.0    5
           ..
199200.0    1
138700.0    1
172700.0    1
134600.0    1
113100.0    1
Name: count, Length: 917, dtype: int64

In [100]:
df_clean["AveragePurchaseAmount"] = df_clean["AveragePurchaseAmount"].astype("string")

In [101]:
df_clean["AveragePurchaseAmount"] = df_clean["AveragePurchaseAmount"].str.replace(",", "", regex=False)

In [102]:
df_clean["AveragePurchaseAmount"].value_counts(dropna=False)

AveragePurchaseAmount
5000.0      9
143400.0    7
148300.0    5
99600.0     5
116900.0    5
           ..
199200.0    1
138700.0    1
172700.0    1
134600.0    1
113100.0    1
Name: count, Length: 917, dtype: Int64

In [103]:
df_clean["AveragePurchaseAmount"] = df_clean["AveragePurchaseAmount"].str.strip()

In [104]:
df_clean["AveragePurchaseAmount"].value_counts(dropna=False)

AveragePurchaseAmount
5000.0      9
143400.0    7
148300.0    5
99600.0     5
116900.0    5
           ..
199200.0    1
138700.0    1
172700.0    1
134600.0    1
113100.0    1
Name: count, Length: 917, dtype: Int64

In [105]:
df_clean["AveragePurchaseAmount"] = pd.to_numeric(df_clean["AveragePurchaseAmount"], errors="coerce")

In [106]:
df_clean["AveragePurchaseAmount"].value_counts(dropna=False)

AveragePurchaseAmount
5000.0      9
143400.0    7
148300.0    5
99600.0     5
116900.0    5
           ..
199200.0    1
138700.0    1
172700.0    1
134600.0    1
113100.0    1
Name: count, Length: 912, dtype: Int64

In [107]:
df_clean["AveragePurchaseAmount"].dtype

Float64Dtype()

In [108]:
original_present = df["AveragePurchaseAmount"].notna()
converted_missing = df_clean["AveragePurchaseAmount"].isna()
failed = original_present & converted_missing
df.loc[failed, "AveragePurchaseAmount"].value_counts(dropna=False)

Series([], Name: count, dtype: int64)

In [109]:
df_clean[["Age", "AveragePurchaseAmount"]].describe()

,Age,AveragePurchaseAmount
count,1312.000000,1317.0
mean,39.556402,120754.821564
std,11.875974,52098.456169
min,18.000000,5000.0
25%,32.000000,83800.0
50%,39.000000,117500.0
75%,47.000000,151500.0
max,130.000000,280400.0


In [110]:
df_clean["LastPurchaseDate"].head(10)

1     2026-04-22
2     2026-06-06
3     2026-05-07
4     2026-04-16
6     2026-01-31
7     2026-01-21
8     2026-06-12
9     2026-03-19
10    2026-02-18
11    2026-03-31
Name: LastPurchaseDate, dtype: str

In [111]:
df_clean["LastPurchaseDate"] = pd.to_datetime(df_clean["LastPurchaseDate"], errors="coerce", format="mixed")

In [112]:
df_clean["LastPurchaseDate"].head(10)

1    2026-04-22
2    2026-06-06
3    2026-05-07
4    2026-04-16
6    2026-01-31
7    2026-01-21
8    2026-06-12
9    2026-03-19
10   2026-02-18
11   2026-03-31
Name: LastPurchaseDate, dtype: datetime64[us]

In [113]:
df_clean["LastPurchaseDate"].isna().sum()

np.int64(0)

In [114]:
original_present = df["LastPurchaseDate"].notna()
converted_missing = df_clean["LastPurchaseDate"].isna()
failed = original_present & converted_missing
df.loc[failed, "LastPurchaseDate"].value_counts(dropna=False)

Series([], Name: count, dtype: int64)

In [115]:
df_clean["SignupDate"].head(10)

1     2021-07-16
2     2022-02-15
3     2021-10-27
4     2025-11-08
6     2025-01-28
7     2021-07-01
8     2024-05-03
9     2023-10-13
10    2024-09-05
11    2022-06-12
Name: SignupDate, dtype: str

In [116]:
df_clean["SignupDate"] = pd.to_datetime(df_clean["SignupDate"], errors="coerce", format="mixed")

In [117]:
df_clean["SignupDate"].head(10)

1    2021-07-16
2    2022-02-15
3    2021-10-27
4    2025-11-08
6    2025-01-28
7    2021-07-01
8    2024-05-03
9    2023-10-13
10   2024-09-05
11   2022-06-12
Name: SignupDate, dtype: datetime64[us]

In [118]:
df_clean["SignupDate"].isna().sum()

np.int64(0)

In [119]:
original_present = df["SignupDate"].notna()
converted_missing = df_clean["SignupDate"].isna()
failed = original_present & converted_missing
df.loc[failed, "SignupDate"].value_counts(dropna=False)

Series([], Name: count, dtype: int64)

In [120]:
df_clean["Gender"].value_counts(dropna=False)

Gender
Female      599
Male        597
Other        42
female       15
male         14
 MALE        14
 FEMALE      12
F            12
M            12
Name: count, dtype: int64

In [121]:
df_clean["Gender"] = df_clean["Gender"].astype("string")

In [122]:
df_clean["Gender"] = df_clean["Gender"].str.strip()

In [123]:
df_clean["Gender"].value_counts(dropna=False)

Gender
Female    599
Male      597
Other      42
female     15
male       14
MALE       14
FEMALE     12
F          12
M          12
Name: count, dtype: Int64

In [124]:
gender_map = {
    "female": "Female",
    "f": "Female",
    "male": "Male",
    "m": "Male",
    "other": "Other"
}

In [125]:
df_clean["Gender"] = df_clean["Gender"].map(gender_map)

In [126]:
df_clean["Gender"].value_counts(dropna=False)

Gender
NaN       1288
Female      15
Male        14
Name: count, dtype: int64

In [127]:
df_clean["Region"].value_counts(dropna=False)

Region
Seoul         347
Gyeonggi      310
Daejeon       147
Gwangju       135
Incheon       119
Busan         109
Daegu         101
 GYEONGGI      11
 Seoul         11
seoul          10
gyeonggi       10
busan           4
 Busan          3
Name: count, dtype: int64

In [128]:
df_clean["Region"] = df_clean["Region"].astype("string")

In [129]:
df_clean["Region"] = df_clean["Region"].str.strip()

In [130]:
df_clean["Region"].value_counts(dropna=False)

Region
Seoul       358
Gyeonggi    310
Daejeon     147
Gwangju     135
Incheon     119
Busan       112
Daegu       101
GYEONGGI     11
seoul        10
gyeonggi     10
busan         4
Name: count, dtype: Int64

In [131]:
df_clean["Region"] = df_clean["Region"].str.lower()

In [132]:
df_clean["Region"].value_counts(dropna=False)

Region
seoul       368
gyeonggi    331
daejeon     147
gwangju     135
incheon     119
busan       116
daegu       101
Name: count, dtype: Int64

In [133]:
df_clean["Region"] = df_clean["Region"].str.title()

In [134]:
df_clean["Region"].value_counts(dropna=False)

Region
Seoul       368
Gyeonggi    331
Daejeon     147
Gwangju     135
Incheon     119
Busan       116
Daegu       101
Name: count, dtype: Int64

In [135]:
df_clean["MembershipLevel"].value_counts(dropna=False)

MembershipLevel
Basic      571
Silver     354
Gold       228
VIP         93
 Basic      20
basic       16
SILVER      11
silver      11
 Gold        7
gold         6
Name: count, dtype: int64

In [136]:
df_clean["MembershipLevel"] = df_clean["MembershipLevel"].astype("string")

In [137]:
df_clean["MembershipLevel"] = df_clean["MembershipLevel"].str.strip()

In [138]:
df_clean["MembershipLevel"].value_counts(dropna=False)

MembershipLevel
Basic     591
Silver    354
Gold      235
VIP        93
basic      16
SILVER     11
silver     11
gold        6
Name: count, dtype: Int64

In [139]:
df_clean["CouponUsed"].value_counts(dropna=False)

CouponUsed
N      701
Y      536
No      25
 n      21
 y      18
Yes     16
Name: count, dtype: int64

In [140]:
df_clean["CouponUsed"] = df_clean["CouponUsed"].astype("string")

In [141]:
df_clean["CouponUsed"] = df_clean["CouponUsed"].str.strip()

In [142]:
df_clean["CouponUsed"].value_counts(dropna=False)

CouponUsed
N      701
Y      536
No      25
n       21
y       18
Yes     16
Name: count, dtype: Int64

In [143]:
df_clean["CouponUsed"] = df_clean["CouponUsed"].str.lower()

In [144]:
df_clean["CouponUsed"].value_counts(dropna=False)

CouponUsed
n      722
y      554
no      25
yes     16
Name: count, dtype: Int64

In [145]:
coupon_map = {"y": "Y", "yes": "Y", "n": "N", "no": "N"}

In [146]:
df_clean["CouponUsed"] = df_clean["CouponUsed"].map(coupon_map)

In [147]:
df_clean["CouponUsed"].value_counts(dropna=False)

CouponUsed
N    747
Y    570
Name: count, dtype: int64

In [148]:
df_clean["PreferredCategory"].value_counts(dropna=False)

PreferredCategory
Fashion          301
Electronics      224
Food             218
Beauty           206
Home             182
Sports           146
fashion            9
FASHION            8
 Beauty            6
 Electronics       6
electronics        6
beauty             5
Name: count, dtype: int64

In [149]:
df_clean["PreferredCategory"] = df_clean["PreferredCategory"].astype("string")

In [150]:
df_clean["PreferredCategory"] = df_clean["PreferredCategory"].str.strip()

In [151]:
df_clean["PreferredCategory"].value_counts(dropna=False)

PreferredCategory
Fashion        301
Electronics    230
Food           218
Beauty         212
Home           182
Sports         146
fashion          9
FASHION          8
electronics      6
beauty           5
Name: count, dtype: Int64

In [152]:
df_clean["PreferredCategory"] = df_clean["PreferredCategory"].str.lower()

In [153]:
df_clean["PreferredCategory"].value_counts(dropna=False)

PreferredCategory
fashion        318
electronics    236
food           218
beauty         217
home           182
sports         146
Name: count, dtype: Int64

In [154]:
df_clean.dtypes

CustomerID                          str
Gender                              str
Age                             float64
Region                           string
MembershipLevel                  string
VisitCount                        int64
AveragePurchaseAmount           Float64
TotalPurchaseAmount             float64
SatisfactionScore               float64
LastPurchaseDate         datetime64[us]
CouponUsed                          str
PreferredCategory                string
SignupDate               datetime64[us]
Email                               str
PurchaseStatus                    int64
dtype: object

In [155]:
# TODO: 결측치와 전체 행 기준 중복 데이터를 처리하세요.
missing_before_fill = df_clean.isna().sum()
missing_before_fill


# TODO: 처리 전후 결과를 확인하세요.
print("정제 전 행 수:", len(df))
print("정제 후 행 수:", len(df_clean))

정제 전 행 수: 1317
정제 후 행 수: 1317


In [156]:
age_median = df_clean["Age"].median()
print("대체할 중앙값:", age_median)

대체할 중앙값: 39.0


In [157]:
df_clean["Age"] = df_clean["Age"].fillna(age_median)

In [158]:
df_clean["Age"].value_counts(dropna=False)

Age
47.0     51
41.0     50
38.0     49
32.0     49
46.0     48
         ..
99.0      1
101.0     1
68.0      1
67.0      1
64.0      1
Name: count, Length: 62, dtype: int64

In [159]:
df_clean["Age"].isna().sum()

np.int64(0)

In [160]:
average_median = df_clean["AveragePurchaseAmount"].median()
print("대체할 중앙값:", average_median)

대체할 중앙값: 117500.0


In [161]:
df_clean["AveragePurchaseAmount"] = df_clean["AveragePurchaseAmount"].fillna(average_median)

In [162]:
df_clean["AveragePurchaseAmount"].value_counts(dropna=False)

AveragePurchaseAmount
5000.0      9
143400.0    7
148300.0    5
99600.0     5
116900.0    5
           ..
199200.0    1
138700.0    1
172700.0    1
134600.0    1
113100.0    1
Name: count, Length: 912, dtype: Int64

In [163]:
df_clean["AveragePurchaseAmount"].isna().sum()

np.int64(0)

In [164]:
satisfaction_median = df_clean["SatisfactionScore"].median()
print("대체할 중앙값:", satisfaction_median)

대체할 중앙값: 3.7


In [165]:
df_clean["SatisfactionScore"] = df_clean["SatisfactionScore"].fillna(satisfaction_median)

In [166]:
df_clean["SatisfactionScore"].value_counts(dropna=False)

SatisfactionScore
3.3    70
3.6    70
3.9    64
4.1    62
4.2    61
3.8    61
3.5    60
3.7    58
3.2    58
3.0    55
3.4    54
4.0    51
5.0    50
4.3    48
3.1    46
4.4    45
4.5    44
4.6    36
2.8    35
4.7    34
2.7    31
2.9    31
2.6    29
4.9    29
4.8    28
2.4    23
2.5    19
2.3    18
2.2    14
2.0     9
2.1     8
1.8     5
1.6     3
1.4     3
1.9     2
1.1     1
1.5     1
1.7     1
Name: count, dtype: int64

In [167]:
df_clean["SatisfactionScore"].isna().sum()

np.int64(0)

In [168]:
Gender_modes = df_clean["Gender"].mode()
Gender_modes

0    Female
Name: Gender, dtype: str

In [169]:
Gender_mode = Gender_modes.iloc[0]
print("대체할 최빈값:", Gender_mode)

대체할 최빈값: Female


In [170]:
df_clean["Gender"] = df_clean["Gender"].fillna(Gender_mode)

In [171]:
df_clean["Gender"].value_counts(dropna=False)

Gender
Female    1303
Male        14
Name: count, dtype: int64

In [172]:
Region_modes = df_clean["Region"].mode()
Region_modes

0    Seoul
Name: Region, dtype: string

In [173]:
Region_mode = Region_modes.iloc[0]
print("대체할 최빈값:", Region_mode)

대체할 최빈값: Seoul


In [174]:
df_clean["Region"] = df_clean["Region"].fillna(Region_mode)

In [175]:
df_clean["Region"].value_counts(dropna=False)

Region
Seoul       368
Gyeonggi    331
Daejeon     147
Gwangju     135
Incheon     119
Busan       116
Daegu       101
Name: count, dtype: Int64

In [176]:
MembershipLevel_modes = df_clean["MembershipLevel"].mode()
MembershipLevel_modes

0    Basic
Name: MembershipLevel, dtype: string

In [177]:
MembershipLevel_mode = MembershipLevel_modes.iloc[0]
print("대체할 최빈값:", MembershipLevel_mode)

대체할 최빈값: Basic


In [178]:
df_clean["MembershipLevel"] = df_clean["MembershipLevel"].fillna(MembershipLevel_mode)

In [179]:
df_clean["MembershipLevel"].value_counts(dropna=False)

MembershipLevel
Basic     591
Silver    354
Gold      235
VIP        93
basic      16
SILVER     11
silver     11
gold        6
Name: count, dtype: Int64

In [180]:
CouponUsed_modes = df_clean["CouponUsed"].mode()
CouponUsed_modes

0    N
Name: CouponUsed, dtype: str

In [181]:
CouponUsed_mode = CouponUsed_modes.iloc[0]
print("대체할 최빈값:", CouponUsed_mode)

대체할 최빈값: N


In [182]:
df_clean["CouponUsed"] = df_clean["CouponUsed"].fillna(CouponUsed_mode)

In [183]:
df_clean["CouponUsed"].value_counts(dropna=False)

CouponUsed
N    747
Y    570
Name: count, dtype: int64

In [184]:
PreferredCategory_modes = df_clean["PreferredCategory"].mode()
PreferredCategory_modes

0    fashion
Name: PreferredCategory, dtype: string

In [185]:
PreferredCategory_mode = PreferredCategory_modes.iloc[0]
print("대체할 최빈값:", PreferredCategory_mode)

대체할 최빈값: fashion


In [186]:
df_clean["PreferredCategory"] = df_clean["PreferredCategory"].fillna(PreferredCategory_mode)

In [187]:
df_clean["PreferredCategory"].value_counts(dropna=False)

PreferredCategory
fashion        318
electronics    236
food           218
beauty         217
home           182
sports         146
Name: count, dtype: Int64

In [188]:
valid_dates = df_clean["LastPurchaseDate"].dropna()
valid_dates = valid_dates.sort_values()
valid_dates.head()

707    2025-08-15
1242   2025-08-26
161    2025-08-29
1084   2025-08-31
1208   2025-09-11
Name: LastPurchaseDate, dtype: datetime64[us]

In [189]:
middle_position = len(valid_dates) // 2
LastPurchaseDate_middle = valid_dates.iloc[middle_position]
print("대체할 날짜:", LastPurchaseDate_middle)

대체할 날짜: 2026-04-24 00:00:00


In [190]:
df_clean["LastPurchaseDate"] = df_clean["LastPurchaseDate"].fillna(LastPurchaseDate_middle)

In [192]:
df_clean["LastPurchaseDate"].isna().sum()

np.int64(0)

In [193]:
valid_dates = df_clean["SignupDate"].dropna()
valid_dates = valid_dates.sort_values()
valid_dates.head()

538    2020-06-26
1442   2020-06-28
796    2020-06-29
598    2020-06-29
299    2020-06-30
Name: SignupDate, dtype: datetime64[us]

In [194]:
middle_position = len(valid_dates) // 2
SignupDate_middle = valid_dates.iloc[middle_position]
print("대체할 날짜:", SignupDate_middle)

대체할 날짜: 2023-04-22 00:00:00


In [195]:
df_clean["SignupDate"] = df_clean["SignupDate"].fillna(SignupDate_middle)

In [196]:
df_clean["SignupDate"].isna().sum()

np.int64(0)

In [197]:
print("제거 전 행 수:", len(df_clean))
print("제거할 중복 수:", df_clean.duplicated().sum())

제거 전 행 수: 1317
제거할 중복 수: 54


In [198]:
df_clean = df_clean.drop_duplicates()
df_clean = df_clean.reset_index(drop=True)

In [199]:
print("제거 후 행 수:", len(df_clean))
df_clean.isna().sum()

제거 후 행 수: 1263


CustomerID               0
Gender                   0
Age                      0
Region                   0
MembershipLevel          0
VisitCount               0
AveragePurchaseAmount    0
TotalPurchaseAmount      0
SatisfactionScore        0
LastPurchaseDate         0
CouponUsed               0
PreferredCategory        0
SignupDate               0
Email                    0
PurchaseStatus           0
dtype: int64

In [200]:

# TODO: Q1, Q3, IQR, 하한과 상한을 반환하세요.
age_q1 = df_clean["Age"].quantile(0.25)
age_q3 = df_clean["Age"].quantile(0.75)
age_iqr = age_q3 - age_q1
print("Q1:", age_q1)
print("Q3:", age_q3)
print("IQR:", age_iqr)

age_lower = age_q1 - 1.5 * age_iqr
age_upper = age_q3 + 1.5 * age_iqr
print("하한:", age_lower)
print("상한:", age_upper)

age_outliers = (df_clean["Age"] < age_lower) | (df_clean["Age"] > age_upper)
age_before_count = age_outliers.sum()
print("이상치 후보 수:", age_before_count)
df_clean.loc[age_outliers, ["CustomerID", "Age"]].head()


Q1: 32.0
Q3: 47.0
IQR: 15.0
하한: 9.5
상한: 69.5
이상치 후보 수: 11


,CustomerID,Age
1,CUST100168,105.0
445,CUST100069,72.0
472,CUST100234,79.0
486,CUST100496,120.0
503,CUST101281,76.0


In [201]:
visit_q1 = df_clean["VisitCount"].quantile(0.25)
visit_q3 = df_clean["VisitCount"].quantile(0.75)
visit_iqr = visit_q3 - visit_q1
print("Q1:", visit_q1)
print("Q3:", visit_q3)
print("IQR:", visit_iqr)

visit_lower = visit_q1 - 1.5 * visit_iqr
visit_upper = visit_q3 + 1.5 * visit_iqr
print("하한:", visit_lower)
print("상한:", visit_upper)

visit_outliers = (df_clean["VisitCount"] < visit_lower) | (df_clean["VisitCount"] > visit_upper)
visit_before_count = visit_outliers.sum()
print("이상치 후보 수:", visit_before_count)
df_clean.loc[visit_outliers, ["CustomerID", "VisitCount"]].head()

Q1: 7.0
Q3: 11.0
IQR: 4.0
하한: 1.0
상한: 17.0
이상치 후보 수: 12


,CustomerID,VisitCount
87,CUST100455,170
162,CUST100565,21
242,CUST101188,19
333,CUST101168,19
335,CUST101185,130


In [204]:

Q1 = df_clean["AveragePurchaseAmount"].quantile(0.25)
Q3 = df_clean["AveragePurchaseAmount"].quantile(0.75)

average_iqr = Q3 - Q1

print("IQR:", average_iqr)



IQR: 68450.0


In [208]:
average_Q1 = df_clean["AveragePurchaseAmount"].quantile(0.25)
average_Q3 = df_clean["AveragePurchaseAmount"].quantile(0.75)
average_iqr = average_Q3 - average_Q1

print("Q1:", average_Q1)
print("Q3:", average_Q3)
print("IQR:", average_iqr)


average_lower = average_Q1 - 1.5 * average_iqr
average_upper = average_Q3 + 1.5 * average_iqr
print("하한:", average_lower)
print("상한:", average_upper)

average_outliers = (df_clean["AveragePurchaseAmount"] < average_lower) | (df_clean["AveragePurchaseAmount"] > average_upper)
average_before_count = average_outliers.sum()
print("이상치 후보 수:", average_before_count)
df_clean.loc[average_outliers, ["CustomerID", "AveragePurchaseAmount"]].head()



Q1: 83650.0
Q3: 152100.0
IQR: 68450.0
하한: -19025.0
상한: 254775.0
이상치 후보 수: 13


,CustomerID,AveragePurchaseAmount
4,CUST100600,272600.0
49,CUST100848,280400.0
95,CUST101043,255100.0
143,CUST100068,258300.0
232,CUST100347,277700.0


In [209]:
total_q1 = df_clean["TotalPurchaseAmount"].quantile(0.25)
total_q3 = df_clean["TotalPurchaseAmount"].quantile(0.75)
total_iqr = total_q3 - total_q1
print("Q1:", total_q1)
print("Q3:", total_q3)
print("IQR:", total_iqr)

Q1: 186350.0
Q3: 542200.0
IQR: 355850.0


In [210]:
total_lower = total_q1 - 1.5 * total_iqr
total_upper = total_q3 + 1.5 * total_iqr
print("하한:", total_lower)
print("상한:", total_upper)

하한: -347425.0
상한: 1075975.0


In [211]:
total_outliers = (df_clean["TotalPurchaseAmount"] < total_lower) | (df_clean["TotalPurchaseAmount"] > total_upper)
total_before_count = total_outliers.sum()
print("이상치 후보 수:", total_before_count)
df_clean.loc[total_outliers, ["CustomerID", "TotalPurchaseAmount"]].head()

이상치 후보 수: 45


,CustomerID,TotalPurchaseAmount
4,CUST100600,1164900.0
42,CUST100978,1120600.0
49,CUST100848,1593800.0
80,CUST100791,1334100.0
90,CUST100014,21000000.0


In [212]:
satisfaction_q1 = df_clean["SatisfactionScore"].quantile(0.25)
satisfaction_q3 = df_clean["SatisfactionScore"].quantile(0.75)
satisfaction_iqr = satisfaction_q3 - satisfaction_q1
print("Q1:", satisfaction_q1)
print("Q3:", satisfaction_q3)
print("IQR:", satisfaction_iqr)

Q1: 3.1
Q3: 4.2
IQR: 1.1


In [213]:
satisfaction_lower = satisfaction_q1 - 1.5 * satisfaction_iqr
satisfaction_upper = satisfaction_q3 + 1.5 * satisfaction_iqr
print("하한:", satisfaction_lower)
print("상한:", satisfaction_upper)

하한: 1.45
상한: 5.8500000000000005


In [214]:
satisfaction_outliers = (df_clean["SatisfactionScore"] < satisfaction_lower) | (df_clean["SatisfactionScore"] > satisfaction_upper)
satisfaction_before_count = satisfaction_outliers.sum()
print("이상치 후보 수:", satisfaction_before_count)
df_clean.loc[satisfaction_outliers, ["CustomerID", "SatisfactionScore"]].head()

이상치 후보 수: 4


,CustomerID,SatisfactionScore
150,CUST101157,1.1
650,CUST100105,1.4
1025,CUST101478,1.4
1260,CUST100737,1.4


In [215]:
df_before_outlier = df_clean.copy()

In [216]:
df_clean["Age"] = df_clean["Age"].astype("float64")
df_clean["Age"] = df_clean["Age"].clip(lower=age_lower, upper=age_upper)

In [217]:
changed = df_before_outlier["Age"] != df_clean["Age"]
print("변경한 값 수:", changed.sum())
comparison = pd.DataFrame()
comparison["처리 전"] = df_before_outlier.loc[changed, "Age"]
comparison["처리 후"] = df_clean.loc[changed, "Age"]
comparison.head()

변경한 값 수: 11


,처리 전,처리 후
1,105.0,69.5
445,72.0,69.5
472,79.0,69.5
486,120.0,69.5
503,76.0,69.5


In [218]:
df_clean["Age"].value_counts(dropna=False)

Age
47.0    49
41.0    49
38.0    49
32.0    47
46.0    46
36.0    46
39.0    44
37.0    44
43.0    42
35.0    42
44.0    41
31.0    40
42.0    39
18.0    39
33.0    39
40.0    36
49.0    35
48.0    35
34.0    34
45.0    34
30.0    30
29.0    29
25.0    27
50.0    26
27.0    24
53.0    23
28.0    23
26.0    22
54.0    19
51.0    19
23.0    17
52.0    17
21.0    17
57.0    15
56.0    13
24.0    12
69.5    11
22.0    10
60.0    10
55.0    10
19.0    10
59.0     9
20.0     7
63.0     6
61.0     6
58.0     6
62.0     4
65.0     4
66.0     3
69.0     1
68.0     1
67.0     1
64.0     1
Name: count, dtype: int64

In [219]:
df_clean["VisitCount"] = df_clean["VisitCount"].astype("float64")
df_clean["VisitCount"] = df_clean["VisitCount"].clip(lower=visit_lower, upper=visit_upper)

In [220]:
changed = df_before_outlier["VisitCount"] != df_clean["VisitCount"]
print("변경한 값 수:", changed.sum())
comparison = pd.DataFrame()
comparison["처리 전"] = df_before_outlier.loc[changed, "VisitCount"]
comparison["처리 후"] = df_clean.loc[changed, "VisitCount"]
comparison.head()

변경한 값 수: 12


,처리 전,처리 후
87,170,17.0
162,21,17.0
242,19,17.0
333,19,17.0
335,130,17.0


In [221]:
df_clean["VisitCount"].value_counts(dropna=False)

VisitCount
9.0     168
8.0     156
7.0     155
10.0    152
11.0    133
6.0     121
12.0     92
5.0      74
13.0     57
4.0      40
14.0     39
15.0     27
17.0     17
3.0      16
16.0     13
2.0       3
Name: count, dtype: int64

In [222]:
df_clean["AveragePurchaseAmount"] = df_clean["AveragePurchaseAmount"].astype("float64")
df_clean["AveragePurchaseAmount"] = df_clean["AveragePurchaseAmount"].clip(lower=average_lower, upper=average_upper)

In [223]:
changed = df_before_outlier["AveragePurchaseAmount"] != df_clean["AveragePurchaseAmount"]
print("변경한 값 수:", changed.sum())
comparison = pd.DataFrame()
comparison["처리 전"] = df_before_outlier.loc[changed, "AveragePurchaseAmount"]
comparison["처리 후"] = df_clean.loc[changed, "AveragePurchaseAmount"]
comparison.head()

변경한 값 수: 13


,처리 전,처리 후
4,272600.0,254775.0
49,280400.0,254775.0
95,255100.0,254775.0
143,258300.0,254775.0
232,277700.0,254775.0


In [224]:
df_clean["AveragePurchaseAmount"].value_counts(dropna=False)

AveragePurchaseAmount
254775.0    13
5000.0       9
116900.0     5
143400.0     5
148300.0     4
            ..
199200.0     1
138700.0     1
172700.0     1
134600.0     1
113100.0     1
Name: count, Length: 900, dtype: int64

In [225]:
df_clean["TotalPurchaseAmount"] = df_clean["TotalPurchaseAmount"].astype("float64")
df_clean["TotalPurchaseAmount"] = df_clean["TotalPurchaseAmount"].clip(lower=total_lower, upper=total_upper)

In [226]:
changed = df_before_outlier["TotalPurchaseAmount"] != df_clean["TotalPurchaseAmount"]
print("변경한 값 수:", changed.sum())
comparison = pd.DataFrame()
comparison["처리 전"] = df_before_outlier.loc[changed, "TotalPurchaseAmount"]
comparison["처리 후"] = df_clean.loc[changed, "TotalPurchaseAmount"]
comparison.head()

변경한 값 수: 45


,처리 전,처리 후
4,1164900.0,1075975.0
42,1120600.0,1075975.0
49,1593800.0,1075975.0
80,1334100.0,1075975.0
90,21000000.0,1075975.0


In [227]:
df_clean["TotalPurchaseAmount"].value_counts(dropna=False)

TotalPurchaseAmount
1075975.0    45
408100.0      3
97400.0       3
227000.0      3
358000.0      3
             ..
59700.0       1
417000.0      1
799600.0      1
92800.0       1
486200.0      1
Name: count, Length: 1124, dtype: int64

In [228]:
df_clean["SatisfactionScore"] = df_clean["SatisfactionScore"].clip(lower=1, upper=5)

In [229]:
changed = df_before_outlier["SatisfactionScore"] != df_clean["SatisfactionScore"]
print("변경한 값 수:", changed.sum())
comparison = pd.DataFrame()
comparison["처리 전"] = df_before_outlier.loc[changed, "SatisfactionScore"]
comparison["처리 후"] = df_clean.loc[changed, "SatisfactionScore"]
comparison.head()

변경한 값 수: 0


,처리 전,처리 후


In [230]:
df_clean["SatisfactionScore"].value_counts(dropna=False)

SatisfactionScore
3.6    67
3.3    66
3.9    63
4.2    59
3.8    58
3.5    58
4.1    56
3.7    56
3.2    56
3.4    53
3.0    51
4.0    50
5.0    48
3.1    46
4.3    46
4.4    44
4.5    43
4.6    36
2.8    33
2.9    30
4.7    30
2.7    29
2.6    28
4.9    28
4.8    25
2.4    23
2.5    18
2.3    17
2.2    13
2.0     9
2.1     8
1.8     5
1.6     3
1.4     3
1.9     2
1.1     1
1.5     1
1.7     1
Name: count, dtype: int64

In [231]:
quality_report = pd.DataFrame()
quality_report["원본 결측 수"] = df.isna().sum()
quality_report["정제 후 결측 수"] = df_clean.isna().sum()
quality_report

,원본 결측 수,정제 후 결측 수
CustomerID,0,0
Gender,0,0
Age,0,0
Region,0,0
MembershipLevel,0,0
VisitCount,0,0
AveragePurchaseAmount,0,0
TotalPurchaseAmount,0,0
SatisfactionScore,0,0
LastPurchaseDate,0,0


In [232]:
print("행 수:", len(df), "→", len(df_clean))
print("전체 행 중복:", df.duplicated().sum(), "→", df_clean.duplicated().sum())

행 수: 1317 → 1263
전체 행 중복: 54 → 0


In [233]:
q1 = df_clean["Age"].quantile(0.25)
q3 = df_clean["Age"].quantile(0.75)
iqr = q3 - q1
lower = q1 - 1.5 * iqr
upper = q3 + 1.5 * iqr
outliers = (df_clean["Age"] < lower) | (df_clean["Age"] > upper)
print("처리 전 후보 수:", age_before_count)
print("처리 후 후보 수:", outliers.sum())

처리 전 후보 수: 11
처리 후 후보 수: 0


In [234]:
q1 = df_clean["VisitCount"].quantile(0.25)
q3 = df_clean["VisitCount"].quantile(0.75)
iqr = q3 - q1
lower = q1 - 1.5 * iqr
upper = q3 + 1.5 * iqr
outliers = (df_clean["VisitCount"] < lower) | (df_clean["VisitCount"] > upper)
print("처리 전 후보 수:", visit_before_count)
print("처리 후 후보 수:", outliers.sum())

처리 전 후보 수: 12
처리 후 후보 수: 0


In [235]:
q1 = df_clean["AveragePurchaseAmount"].quantile(0.25)
q3 = df_clean["AveragePurchaseAmount"].quantile(0.75)
iqr = q3 - q1
lower = q1 - 1.5 * iqr
upper = q3 + 1.5 * iqr
outliers = (df_clean["AveragePurchaseAmount"] < lower) | (df_clean["AveragePurchaseAmount"] > upper)
print("처리 전 후보 수:", average_before_count)
print("처리 후 후보 수:", outliers.sum())

처리 전 후보 수: 13
처리 후 후보 수: 0


In [236]:
q1 = df_clean["TotalPurchaseAmount"].quantile(0.25)
q3 = df_clean["TotalPurchaseAmount"].quantile(0.75)
iqr = q3 - q1
lower = q1 - 1.5 * iqr
upper = q3 + 1.5 * iqr
outliers = (df_clean["TotalPurchaseAmount"] < lower) | (df_clean["TotalPurchaseAmount"] > upper)
print("처리 전 후보 수:", total_before_count)
print("처리 후 후보 수:", outliers.sum())

처리 전 후보 수: 45
처리 후 후보 수: 0


In [237]:
q1 = df_clean["SatisfactionScore"].quantile(0.25)
q3 = df_clean["SatisfactionScore"].quantile(0.75)
iqr = q3 - q1
lower = q1 - 1.5 * iqr
upper = q3 + 1.5 * iqr
outliers = (df_clean["SatisfactionScore"] < lower) | (df_clean["SatisfactionScore"] > upper)
print("처리 전 후보 수:", satisfaction_before_count)
print("처리 후 후보 수:", outliers.sum())

처리 전 후보 수: 4
처리 후 후보 수: 4


## 문제 3. 분석용 변수 생성 및 다음 단계 데이터 준비

### 업무 상황

정제된 데이터를 CRM팀의 고객군 분석과 구매 예측 업무에 활용하려면, 원본 컬럼만으로 확인하기 어려운 고객 특성을 분석용 변수로 표현해야 합니다.  
필요한 고객을 조건에 따라 추출하고, 이후 Part 2와 Part 3에서 활용할 파생변수를 만든 뒤 최종 데이터를 저장하세요.

### 요구사항

1. 정제된 데이터에서 업무적으로 의미 있는 조건을 설정하여 고객 데이터를 추출하세요.
2. 고객 분석에 활용할 파생변수를 **2개 이상** 생성하세요.
3. 생성한 파생변수의 값과 분포가 적절한지 확인하세요.
4. 최종 데이터의 결측치, 중복값, 고객 식별자와 타깃 값을 검증하세요.
5. 전처리 완료 데이터를 `전자상거래_고객구매_전처리완료.csv`로 저장하세요.

### 힌트

- 연령대, 구매 수준, 최근 구매 여부, 우수 고객 여부 등을 파생변수 후보로 고려할 수 있습니다.
- 조건 추출 결과는 별도 DataFrame으로 확인해도 되며, 최종 데이터 전체를 삭제할 필요는 없습니다.

- 고객 연령을 구간화한 `AgeGroup`을 생성하세요.
- 총구매금액을 기준으로 `PurchaseGrade`를 생성하세요.

In [ ]:

# TODO: 업무적으로 의미 있는 조건을 설정하여 고객 데이터를 추출하세요.

visit_median = df_clean["VisitCount"].median()
visit_median

visit_condition = df_clean["VisitCount"] >= visit_median
purchase_condition = df_clean["TotalPurchaseAmount"] > 0
selected_customers = df_clean[visit_condition & purchase_condition]
selected_customers = selected_customers.copy()

print("조건에 맞는 고객 수:", len(selected_customers))
selected_customers.head()


# TODO: 분석용 파생변수를 2개 이상 생성하세요.
# Age Group
age_bins = [0, 29, 39, 49, 59, float("inf")]
age_labels = ["20대 이하", "30대", "40대", "50대", "60대 이상"]

df_clean["AgeGroup"] = pd.cut(
    df_clean["Age"],
    bins=age_bins,
    labels=age_labels,
    include_lowest=True
)

display(df_clean[["Age", "AgeGroup"]].head(10))


# PurchaseGrade 구매등급 만들기
purchase_q1 = df_clean["TotalPurchaseAmount"].quantile(0.25)
purchase_q3 = df_clean["TotalPurchaseAmount"].quantile(0.75)
print("Q1:", purchase_q1)
print("Q3:", purchase_q3)

df_clean["PurchaseGrade"] = "Medium"
df_clean["PurchaseGrade"].value_counts(dropna=False)


low_condition = df_clean["TotalPurchaseAmount"] <= purchase_q1
df_clean.loc[low_condition, "PurchaseGrade"] = "Low"
df_clean["PurchaseGrade"].value_counts(dropna=False)

high_condition = df_clean["TotalPurchaseAmount"] > purchase_q3
df_clean.loc[high_condition, "PurchaseGrade"] = "High"
df_clean["PurchaseGrade"].value_counts(dropna=False)
df_clean[["TotalPurchaseAmount", "PurchaseGrade"]].head(10)


# TODO: 파생변수의 값과 분포를 확인하세요.
df_clean["AgeGroup"].value_counts(dropna=False)
df_clean["AgeGroup"].isna().sum()
df_clean["PurchaseGrade"].value_counts(dropna=False)
df_clean["PurchaseGrade"].isna().sum()

df_clean.groupby("PurchaseGrade")["TotalPurchaseAmount"].min()
df_clean.groupby("PurchaseGrade")["TotalPurchaseAmount"].max()
df_clean.groupby("PurchaseGrade")["TotalPurchaseAmount"].median()

example_ages = pd.DataFrame()
example_ages["예시 나이"] = [19, 20, 29, 30, 39, 40, 59, 60]
example_ages["구간"] = pd.cut(example_ages["예시 나이"], bins=age_bins, labels=age_labels, include_lowest=True)
example_ages

middle_amount = (purchase_q1 + purchase_q3) / 2
example_amounts = pd.DataFrame()
example_amounts["예시 금액"] = [purchase_q1, middle_amount, purchase_q3, purchase_q3 + 1]
example_amounts["구매등급"] = "Medium"
low_example = example_amounts["예시 금액"] <= purchase_q1
example_amounts.loc[low_example, "구매등급"] = "Low"
high_example = example_amounts["예시 금액"] > purchase_q3
example_amounts.loc[high_example, "구매등급"] = "High"
example_amounts

조건에 맞는 고객 수: 698


,Age,AgeGroup
0,39.0,30대
1,69.5,60대 이상
2,44.0,40대
3,37.0,30대
4,47.0,40대
5,26.0,20대 이하
6,56.0,50대
7,42.0,40대
8,53.0,50대
9,23.0,20대 이하


Q1: 186350.0
Q3: 542200.0


,예시 금액,구매등급
0,186350.0,Low
1,364275.0,Medium
2,542200.0,Medium
3,542201.0,High


In [243]:
purchase_q1 = df_clean["TotalPurchaseAmount"].quantile(0.25)
purchase_q3 = df_clean["TotalPurchaseAmount"].quantile(0.75)
print("Q1:", purchase_q1)
print("Q3:", purchase_q3)

df_clean["PurchaseGrade"] = "Medium"
df_clean["PurchaseGrade"].value_counts(dropna=False)


low_condition = df_clean["TotalPurchaseAmount"] <= purchase_q1
df_clean.loc[low_condition, "PurchaseGrade"] = "Low"
df_clean["PurchaseGrade"].value_counts(dropna=False)

high_condition = df_clean["TotalPurchaseAmount"] > purchase_q3
df_clean.loc[high_condition, "PurchaseGrade"] = "High"
df_clean["PurchaseGrade"].value_counts(dropna=False)
df_clean[["TotalPurchaseAmount", "PurchaseGrade"]].head(10)

Q1: 186350.0
Q3: 542200.0


,TotalPurchaseAmount,PurchaseGrade
0,159200.0,Low
1,1055800.0,High
2,413200.0,Medium
3,576000.0,High
4,1075975.0,High
5,40700.0,Low
6,443500.0,Medium
7,70700.0,Low
8,629800.0,High
9,451500.0,Medium


In [242]:

# TODO: 최종 데이터 품질을 검증하세요.
df_clean.isna().sum()

missing_count = df_clean.isna().sum().sum()
print("전체 결측 수 (0이어야 함):", missing_count)

duplicate_count = df_clean.duplicated().sum()
print("전체 행 중복 수 (0이어야 함):", duplicate_count)

id_duplicate_count = df_clean["CustomerID"].duplicated().sum()
id_missing_count = df_clean["CustomerID"].isna().sum()
print("ID 중복 수:", id_duplicate_count)
print("ID 결측 수:", id_missing_count)

id_text = df_clean["CustomerID"].astype("string")
id_text = id_text.str.strip()
id_empty_count = (id_text == "").sum()
print("공백만 있는 ID 수 (0이어야 함):", id_empty_count)

df_clean["PurchaseStatus"].value_counts(dropna=False)
target_valid = df_clean["PurchaseStatus"].isin([0, 1]).all()
print("타깃이 모두 0 또는 1인가 (True여야 함):", target_valid)

print("AgeGroup" in df_clean.columns)
print("PurchaseGrade" in df_clean.columns)

# TODO: 전처리 완료 데이터를 CSV 파일로 저장하세요.
df_to_save = df_clean.copy()
df_to_save["LastPurchaseDate"] = df_to_save["LastPurchaseDate"].dt.strftime("%Y-%m-%d")
df_to_save["SignupDate"] = df_to_save["SignupDate"].dt.strftime("%Y-%m-%d")
df_to_save.to_csv("전자상거래_고객구매_전처리완료.csv", index=False, encoding="utf-8-sig")

saved_df = pd.read_csv("전자상거래_고객구매_전처리완료.csv")
saved_df.head()
print("정제 데이터 크기:", df_clean.shape)
print("다시 읽은 크기:", saved_df.shape)
print("컬럼 일치:", saved_df.columns.equals(df_clean.columns))
saved_df.isna().sum()

assert saved_df.shape == df_clean.shape
assert saved_df.columns.equals(df_clean.columns)
assert saved_df.isna().sum().sum() == 0
assert saved_df["CustomerID"].is_unique
print("CSV 저장 및 다시 읽기 확인 완료")

전체 결측 수 (0이어야 함): 0
전체 행 중복 수 (0이어야 함): 0
ID 중복 수: 0
ID 결측 수: 0
공백만 있는 ID 수 (0이어야 함): 0
타깃이 모두 0 또는 1인가 (True여야 함): True
True
True
정제 데이터 크기: (1263, 17)
다시 읽은 크기: (1263, 17)
컬럼 일치: True
CSV 저장 및 다시 읽기 확인 완료


# 문제 4. 데이터 정제 및 파생변수 설계 근거 설명

## 업무 상황

CRM팀은 전처리 결과가 단순히 실행되는 것뿐 아니라,
적용한 처리 기준과 파생변수가 실제 분석 목적에 적합한지 확인하려고 합니다.

문제 1~3에서 수행한 결과를 바탕으로 다음 내용을 설명하세요.

## 요구사항

### 4-1. 데이터 품질 문제의 영향과 한계

1. 문제 1에서 확인한 주요 품질 문제를 한 가지 이상 선택하세요.
2. 해당 문제가 이후 EDA 또는 모델링 결과에 미칠 수 있는 영향을 설명하세요.
3. 현재 데이터만으로 판단하기 어려운 점이나 추가 확인이 필요한 조건을 작성하세요.

### 4-2. 결측치·이상치 처리 방법의 선택 근거

1. 적용한 결측치 처리 방법을 한 가지 이상 제시하세요.
2. 해당 방법을 선택한 이유를 데이터 특성과 연결하여 작성하세요.
3. 적용한 이상치 처리 방법과 처리 시 주의할 점을 설명하세요.

### 4-3. 파생변수의 활용 의도와 한계

1. `파생변수1`과 `파생변수2`의 생성 기준을 설명하세요.
2. 각 파생변수가 Part 2 고객 특성 분석에서 어떻게 활용될 수 있는지 작성하세요.
3. 파생변수 사용 시 발생할 수 있는 정보 손실 또는 해석상의 한계를 설명하세요.

## 문제 4 결과 작성

### 4-1. 데이터 품질 문제의 영향과 한계

- 선택한 품질 문제: 결측치와 이상치가 존재하여 결측치가 있는 행을 삭제하여 처리했다. Age, VisitCount, AveragePurchaseAmount 등이 처음에는 문자열로 저장되어 있어 수치형으로 변환하였다.

- EDA 또는 모델링에 미치는 영향: 결측치가 있는 데이터를 사용하면 결과값이 왜곡될 수 있어 모델링 과정에서도 예측 성능에 영향을 줄 수 있다.

- 추가 확인이 필요한 조건 또는 한계: 결측치가 발생한 원인이 명확하지 않다. 또한 이상치로 판단된 값도 실제 오류인지, 구매 금액이 높은 정상적인 고객인지 판단하기 어렵다. 따라서 결측치 발생 원인, 이상치의 실제 거래 여부, 데이터 수집 기간 등을 확인해야한다.

### 4-2. 결측치·이상치 처리 방법의 선택 근거

- 결측치 처리 방법: 결측치가 포함된 행을 삭제했다.(dropna())
- 해당 방법을 선택한 이유: 결측치가 여러 컬럼에 존재했으며 결측값을 처리 안하면 결과 데이터가 왜곡될 수 있기 때문이다.
- 이상치 처리 방법: 수치형 주요 컬럼에 대해 IQR기준을 적용해 이상치를 탐지했고 , 하한과 상한을 벗어나는 값은 제거했다.
- 이상치 처리 시 주의점: 이상치가 반드시 오류 데이터는 아닐 수 있다. 실제로 구매 금액이 높은 고객이거나 방문 횟수가 높은 고객일 수 있기 때문에 확인후 이상치로 처리해야한다.

### 4-3. 파생변수의 활용 의도와 한계

- `AgeGroup` 생성 기준과 활용 목적: AgeGroup은 정제한 Age를 0~29, 29 초과~39, 39 초과~49, 49 초과~59, 59 초과 구간으로 나누고 20대 이하·30대·40대·50대·60대 이상으로 이름 붙였다.
- `PurchaseGrade` 생성 기준과 활용 목적: PurchaseGrade는 정제한 총구매금액의 Q1 186350 이하를 Low, Q1 초과 Q3 542200 이하를 Medium, Q3 초과를 High로 정했다
- 파생변수 사용 시 한계: 구간의 기준에 따라 고객 분류 결과가 달라질 수 있으며, 같은 구간에 속한 고객이라도 실제 구매금액이나 연령 차이가 클 수 있다.